Mcol 4841 dev6 Handle large joins without OOM

davidhall586 opened the pull request:
I moved generateJoinResultSet() and processFE2() into class JoinLocalData. This reduces the calling signature significantly since most of the items it uses for computation exist there.

Commit:
MCOL-4846 dev-6 Handle large join results
Use a loop to shrink the number of results reported per message to something manageable.

Commit:
MCOL-4841 small changes requested by review

Commit:
Add EXTRA threads to prioritythreadpool
prioritythreadpool is configured at startup with a fixed number of threads available. This is to prevent thread thrashing. Since most of the time, BPP job steps are short lived, and a rescheduling mechanism exist if no threads are available, this works to keep cpu wastage to a minimum.

However, if a query or queries consume all the threads in prioritythreadpool and then block (due to the consumer not consuming fast enough) we can run out of threads and no work will be done until some threads unblock. A new mechanism allows for EXTRA threads to be generated for the duration of the blocking action. These threads can act on new queries. When all blocking is completed, these threads will be released when idle.

Commit:
MCOL-4841 dev6 Reconcile with changes in develop-6

Commit:
MCOL-4841 Some format corrections

mariadb-DenisKhalikov commented on dbcon/joblist/diskjoinstep.cpp:
```diff
@@ -484,6 +482,21 @@ void DiskJoinStep::joinFcn()
+                                if (!thjs->getMemory(l_outputRG.getMaxDataSize()))
+                                {
+                                    // calculate guess of size required for error message
+                                    uint64_t memReqd = (unmatched.size() * outputRG.getDataSize(1)) / 1048576;;
```
Typo: extra `;`

mariadb-DenisKhalikov commented on dbcon/joblist/tuple-bps.cpp:
```diff
@@ -2383,10 +2486,7 @@ void TupleBPS::receiveMultiPrimitiveMessages()
             }
 
-            // Join threads.
-            for (uint32_t i = 0, e = fProcessorThreads.size(); i < e; ++i)
```
I see, thanks for the fix =)

mariadb-DenisKhalikov commented on dbcon/joblist/tuple-bps.cpp:
```diff
@@ -2495,9 +2593,8 @@ void TupleBPS::receiveMultiPrimitiveMessages()
-                        processFE2(data->local_outputRG, data->local_fe2Output, data->postJoinRow,
-                                   data->local_fe2OutRow, &rgDatav, &data->local_fe2);
-
+                        data->processFE2(rgDatav);
```
This looks very nice comparing to what it was. Thanks.

mariadb-DenisKhalikov commented on primitives/primproc/batchprimitiveprocessor.cpp:
```diff
@@ -1384,8 +1392,18 @@ void BatchPrimitiveProcessor::executeTupleJoin()
         }
+        // If we've accumulated more than 1048576 (2^20)_ of resultCounts, cut off processing. The caller will
+        // restart to continue where we left off.
+        if (resultCount >= 1048576)
```
Nit: It would be nice to have a named constant for this constant 1048576. And in other places in this code.

drrtuy replied:
+1

davidhall586 replied:
Done

mariadb-DenisKhalikov reviewed (commented):
@dhall-MariaDB thanks for the patch!
The changes for the `TupleBPS` and `tuplehashjoin` LGTM, but for other parts like priority thread pool - I don't have enough understanding for all corner cases, I thinks @drrtuy should take a look on that parts.
Thanks!

drrtuy commented on dbcon/mysql/ha_mcs_impl.cpp:
```diff
@@ -3086,6 +3086,10 @@ void ha_mcs_impl_start_bulk_insert(ha_rows rows, TABLE* table, bool is_cache_ins
             //@bug 6122 Check how many columns have not null constraint. columnn with not null constraint will not show up in header.
             unsigned int numberNotNull = 0;
 
+            // There's a chance that a previously aborted query left garbage in ci->columnTypes
```
IMHO This change is outside the scope of the patch and musn't get into this patch. Moreover the clean-up made in the middle of an operation looks like a side-effect nobody knows about so we better put it where all clean-up happens, e.g. impl_external_lock().

davidhall586 replied:
I disagree. This is something that belongs here. There is a clear() at the tail of the function, but errors may cause early exit, bypassing the clear(). It is common to initialize variables before use. In this case, it's a good safe way to do it.
As for scope, that can be debated, but uninitialized ci->columnTypes caused a bit of pain while testing this patch. Thus this change. To just remove it as not in the scope of the problem will cause unnecessary headache. We'd end up putting it back in at a later time. Or perhaps find every branch that exits the function and add cleanup code? Nah.

drrtuy replied:
No query passes impl_external_lock() and adding uncorrelated side effects makes it hard to debug/maintain the code IMHO. That is why I am strongly against putting it in the middle of  ha_mcs_impl_start_bulk_insert

davidhall586 replied:
I removed the offensive code for now and opened MCOL-4968 to address the issue.

drrtuy commented on oam/etc/Columnstore.xml:
```diff
@@ -96,6 +96,7 @@
 		<!-- <MediumPriorityPercentage>30</MediumPriorityPercentage> -->
 		<!-- <LowPriorityPercentage>10</LowPriorityPercentage> -->
 		<DirectIO>y</DirectIO>
+        <MaxBPPSendQueue>250M</MaxBPPSendQueue>
```
As you know we can't deliver Columnstore.xml changes until we deliver the MCOL-4928.

davidhall586 replied:
This is the default, so we can leave it out of the xml for now.

drrtuy commented on primitives/primproc/bppsendthread.h:
```diff
@@ -124,7 +141,10 @@ class BPPSendThread
+    static uint64_t maxByteSize;
+    // Used to tell the PriorityThreadPool It should consider additional threads because a 
+    // queue full event has happened and a thread has been blocked.
+    boost::shared_ptr<threadpool::PriorityThreadPool> fProcessorPool;
```
Could you elaborate on why do we need smart pointer here?

drrtuy replied:
There can be only one processor pool so it must be safe to use plain PriorityThreadPool pointer.

davidhall586 replied:
It's a shared pointer in PrimitiveServer is why it became a shared pointer in BPPSendThread. Since PrimitiveServer is a singleton whose life, once created, lasts until process termination, and fProcessorPool, is a singleton within it, I don't see why it needs to be a shared pointer.

drrtuy commented on primitives/primproc/bppsendthread.h:
```diff
@@ -124,7 +141,10 @@ class BPPSendThread
     /* secondary queue size restriction based on byte size */
     volatile uint64_t currentByteSize;
-    uint64_t maxByteSize;
+    static uint64_t maxByteSize;
```
Sorry but static member looks a bit hack-ish here.
AFAIK the only place where PP creates BPPSendThread is BPPV ctor(primproc/primitiveserver.cpp) and it is possible to ship the value inside PrimitiveServer which ptr is available in the BPPV ctor.

davidhall586 replied:
Yes indeed we could do that. My intent was to not have to deal with (not spend cpu cycles on) that. It's not horribly expensive, but I am trying to shave where I can.
The drawback of my approach is that a restart is definitely required to notice a change to the xml setting.

drrtuy replied:
I will take another look at this to convince myself static fits here.

davidhall586 replied:
I don't suffer from a prejudice against static members:)

drrtuy replied:
I do b/c it represents a state variable that lives outside the scope of the code that uses it. Nobody guarantees anything about such states. It is hard to maintain, hard to follow, hard to debug.

davidhall586 replied:
So the options are to create a unscoped global, or to spend cpu to unnecessarily re-initialize in the constructor or to call joblist::ResourceManager::instance()->getMaxBPPSendQueue() each time we need the value.

davidhall586 replied:
I find it easier to maintain and follow static than global variables. At least as a static, it's usage is clearly (without an accesor) for use by class members.

davidhall586 replied:
I could add const, if it helps.
If I just remove it from the class declaration, it still works as a global. The reason I put it in the class is for documentation (to make it clear it's use is for and by the class.
Not sure why you find this confusing.

drrtuy replied:
I don't see any reason against moving config access from getMaxBPPSendQueue() code into RM ctor, so that RM::getMaxBPPSendQueue() just reads the value from RM singleton. Finally you can reference RM::getMaxBPPSendQueue() in the BPPSendThread ctor code in bppsendthread.cpp. What do you think about this approach?

davidhall586 replied:
Changed to your recommendation, though I still feel (even with the change to ResourceManager) that a static makes more sense. But for the sake of moving forward, I put it in the sendthread ctor.

drrtuy commented on primitives/primproc/bppsendthread.h:
```diff
@@ -63,7 +63,15 @@ class BPPSendThread
+    bool sizeTooBig()
+    {
+      // keep the queue size below the 100 msg threshold & below the 250MB mark,
+      // but at least 2 msgs so there is always 1 ready to be sent.
```
at least 3 messages according with the condition below.

davidhall586 replied:
changed the comment

drrtuy commented on primitives/primproc/bppsendthread.cpp:
```diff
@@ -53,29 +51,38 @@ BPPSendThread::~BPPSendThread()
+    if (sizeTooBig())
+    {
+        boost::mutex::scoped_lock sl1(respondLock);
+        while (currentByteSize >= maxByteSize && msgQueue.size() > 3 && !die)
```
We have this condition in multiple places. Changing one of them it is easy to forget the others.

drrtuy replied:
Moreover we have the same scope in the parenthesis in two BPPSendThread methods so it is better to wrap it up in a function.

davidhall586 replied:
It could be even harder to read if all this is in a seperate function, though I could do that. However, after a code search. the function sendResults() appears to no longer be used anywhere. I don't know if we would ever want to resurrect it. Perhaps if we just delete the unreferenced function?

davidhall586 replied:
As for the two remaining places where sizeTooBig is called, in sendResult() this is blocking the thread since it's in the middle of a block. In BatchPrimitiveProcessor::operator(), a positive too big will result in the BPP being returned to the pool and the job rescheduled.

drrtuy replied:
I think it is a good idea to comment out/remove unreferenced function.

davidhall586 replied:
I created MCOL-4969 to remove SendResults.

drrtuy commented on primitives/primproc/bppsendthread.cpp:
```diff
@@ -53,29 +51,38 @@ BPPSendThread::~BPPSendThread()
+    // Wait for the queue to empty out a bit if it's stuffed full
+    if (sizeTooBig())
+    {
+        boost::mutex::scoped_lock sl1(respondLock);
```
I personally don't like using boost where std:: elements can be used.

drrtuy replied:
Can you replace boost::mutex with std::mutex?

davidhall586 replied:
NP

drrtuy commented on primitives/primproc/bppsendthread.cpp:
```diff
@@ -53,29 +51,38 @@ BPPSendThread::~BPPSendThread()
+            fProcessorPool->decBlockedThreads();
+            respondWait = false;
+        }
+        sl1.unlock();
```
This call doesn't look necessary b/c scoped_lock dtor releases the mutex exiting this scope.

davidhall586 replied:
It is redundant. removed.

drrtuy commented on utils/threadpool/prioritythreadpool.cpp:
```diff
@@ -158,6 +179,14 @@ void PriorityThreadPool::threadFcn(const Priority preferredQueue) throw()
+                // we don't want this one any more.
+                if (preferredQueue == EXTRA && stopExtra)
+                {
+                    extraThreads--;
```
I am not clear on stopExtra semantics. It looks like this flag signals when there are no blocking threads. If so I can't get why don't we set extraThreads to zero instead of decrementing it?

davidhall586 replied:
extraThreads counts how many extra threads are currently executing. When stopExtra is set, it effectively tells extra threads to go away when finished with whatever they're working on. When a thread goes away, it decrements extraThreads to reflect the current count of extra threads.

If there are blocked threads and extra threads, then this keeps a cap on how many extra threads will be created -- never more than the number of blocked threads.

However, as written, the number of extra threads may grow (if needed) as the number of blocked threads grows. The number of extra threads will not reduce until the number of blocked threads reaches 0 (thus setting stopExtra = true). I consider this to be OK, as blocked threads are a rarity. But if they do happen, they probably happen in bunches.

We may think to set extraThreads to zero in addJob() at the same time we set stopExtra to true. However, there will still be extra threads around until they finish whatever they were working on. During this period, extraThreads would be inaccurate. Meantime If threads block again, there could be confusion. stopExtra will be set to false and new extra threads will be created. There would be more extra threads in existence than extraThreads reflects. This would allow for more extraThreads than blocked threads. While not a killer, it's not what is intended.

drrtuy commented on dbcon/joblist/primitivestep.h:
```diff
@@ -1515,21 +1512,31 @@ class TupleBPS : public BatchPrimitive, public TupleDeliveryStep
                     std::vector<boost::shared_ptr<joiner::TupleJoiner>>& tjoiners, uint32_t smallSideCount,
                     bool doJoin);
 
+      friend class TupleBPS;
```
If you want to make it proper you need to use getter methods for JoinLocalData attributes are accessed in the code. Friend used here gives me an impression that original idea was partially implemented.

davidhall586 replied:
Absolutely true. Friend here is to alleviate having to implement a pile of setter/getters. JoinLocalData was originally a struct as written by Denis and all members were public. I changed it to a class making all the member variables private and added friend so I didn't have to go about the code to change all access to setter/getter. An option is to leave all members public as they were before.

drrtuy replied:
Agree with you on that. The most appropriate way is to add setters/getters but it is up to you whether to stick with this friend approach or to add methods.

drrtuy commented on dbcon/joblist/primitivestep.h:
```diff
@@ -1366,7 +1366,15 @@ class TupleBPS : public BatchPrimitive, public TupleDeliveryStep
+    {
+      return fRm;
+    }
+    bool runFEonPM()
```
Can't we use const specifier for this method?

davidhall586 replied:
const added

drrtuy commented on dbcon/joblist/primitivestep.h:
```diff
@@ -1564,7 +1571,9 @@ class TupleBPS : public BatchPrimitive, public TupleDeliveryStep
       boost::scoped_array<boost::shared_array<uint8_t>> smallNullMemory;
       uint32_t matchCount;
-
+      uint64_t memUsedByOutputRG = 0;
```
Why don't we assign this member in the ctor?

davidhall586 replied:
An argument can be made, but it's moot, since the variable is no longer used. I removed it.

drrtuy commented on dbcon/joblist/distributedenginecomm.cpp:
```diff
@@ -308,13 +308,24 @@ void DistributedEngineComm::Setup()
 
             writeToLog(__FILE__, __LINE__, "Could not connect to PMS" + std::to_string(connectionId) + ": " + ex.what(), LOG_TYPE_ERROR);
             cerr << "Could not connect to PMS" << std::to_string(connectionId) << ": " << ex.what() << endl;
+            
```
We have almost identical exceptions handlers and I suggest you to put the common code into a function parametrizing the messages if needed.

davidhall586 replied:
I thought about this a bit and even started writing the function, but it turns out to not save a lot. Sure I could pass in a ref to newPmCount and the ex.what(). This could standardize the log messages in one place, but I'd still have to test newPmCount ==0 for the break after return. Seems a lot of cpu for very little savings in line count.

drrtuy commented on dbcon/joblist/resourcemanager.h:
```diff
@@ -82,6 +82,7 @@ const uint32_t defaultRequestSize  = 1;
 const uint32_t defaultMaxOutstandingRequests = 20;
 const uint32_t defaultProcessorThreadsPerScan = 16;
 const uint32_t defaultJoinerChunkSize = 16 * 1024 * 1024;
+const uint64_t defaultMaxBPPSendQueue = 250000000; // 250MB
```
Plz put your considerations on this value that you once shared with me in a comment here.

davidhall586 replied:
Done

drrtuy commented on utils/common/atomicops.h:
```diff
@@ -165,6 +165,28 @@ inline bool atomicCAS(volatile T* mem, T comp, T swap)
+template <typename T>
+inline void atomicZero(volatile T* mem)
+{
+#ifdef _MSC_VER
```
I am still sure we won't support windows anymore so it is better to avoid adding new preprocessor Windows related conditions.

davidhall586 replied:
I added this as following the pattern of all the other functions in the lib. Don't assume never. Never tends to byte programmers in the butt.

drrtuy replied:
Hm, all others are removing these blocks but I won't insist on removing this if you don't want.

Commit:
MCOL-4841 dev clean up some things based on review

Commit:
Merge branch 'develop-6' into MCOL-4841-dev6

Commit:
Merge branch 'develop-6' into MCOL-4841-dev6

Commit:
MCOL-4841 dev 6 ExeMgr Crashes after large join
This commit fixes up memory accounting issues in ExeMgr

Commit:
MCOL-4841 remove LDI change
Opened MCOL-4968 to address the issue

Commit:
Merge branch 'develop-6' into MCOL-4841-dev6

drrtuy commented on primitives/primproc/bppsendthread.cpp:
```diff
@@ -254,19 +265,26 @@ void BPPSendThread::mainLoop()
-    sl.unlock();
-    sl2.unlock();
+    {
+        std::lock_guard<std::mutex> sl(msgQueueLock);
```
I was curios about the lifetime of these lock_guards and it seems their dtors are called exiting the scope at line 284 not in the end of the function. [Here](http://cpp.sh/2xqv3) is a fiddler to see and run the code. Is this an expected behavior?

davidhall586 replied:
This is the expected behavior. The lock is only taken if SizeTooBig -- a rare case. It's there to protect the condition variable. Once we're done with the condition, then we don't need or want the lock.

Commit:
MCOL-4841 Add fMaxBPPSendQueue to ResourceManager
This causes the setting to be loaded at run time (requires restart to accept a change) BPPSendthread gets this in it's ctor
Also rolled back changes to TupleHashJoinStep::smallRunnerFcn() that used a local variable to count locally allocated memory, then added it into the global counter at function's end. Not counting the memory globally caused conversion to UM only join way later than it should. This resulted in MCOL-4971.

Commit:
MCOL-4841 make blockedThreads and extraThreads atomic
Also restore previous scope of locks in bppsendthread. There is some small chance the new scope could be incorrect, and the performance boost is negligible. Better safe than sorry.
